src/Monkeypatcher.cc

```diff
@@ -471,18 +493,61 @@ remote_code_ptr Monkeypatcher::get_jump_stub_exit_breakpoint(remote_code_ptr ip,
     return nullptr;
   }
   remote_ptr<uint8_t> pp = ip.to_data_ptr<uint8_t>();
-  auto it = syscallbuf_stubs.upper_bound(pp);
-  if (it == syscallbuf_stubs.begin()) {
+  auto it = syscallbuf_stubs_by_extended_patch.upper_bound(pp);
+  if (it == syscallbuf_stubs_by_extended_patch.begin()) {
     return nullptr;
   }
   --it;
-  auto bp = it->first + it->second.size - it->second.safe_suffix;
-  if (pp == bp || pp == bp - 4) {
-    return remote_code_ptr(bp.as_int());
+  patched_syscall *ps = &syscall_stub_list[it->second];
+  auto bp = it->first + ps->size - ps->safe_suffix;
+  if (pp == bp - 4 || pp == bp - 8) {
```

Comment:
1. I assume we won’t be expecting the exit breakpoint when we are in the bail path? (Otherwise this misses the bail path return)
2. Is the breakpoint address returned here correct? It seems that it is pointing to the data part? (Safe suffix is larger than 8. I assume you meant to use the last instruction which should be - 12 due to the return address I stuffed at the end.

Reply:
(The test failure that required this happens in cont_race about 1-3% of the time iirc.)

Reply:
Ok - I wasn't entirely sure this breakpoint is actually still required, so I'll need to look into that.

Reply:
This breakpoint is needed if we got a signal when we are in the return path of the stub and decided to defer it until the syscallbuf business is done. This would not be needed if we are grantee to not get a signal (comes to think about it I don't think we can grantee that for external signals) or if we just (manually or actually) single step the instructions out. It seems that we may need this on x86 as well now? Or at least a safe suffix so that we can deliver signal when we are back in the stub.